magento/pwa-studio
# [PWA-666]: Update Hastily image optimizer

## Description

While reproducing #2448 we discovered that the image optimizer middleware fails when the underlying proxy that serves images:
- is experiencing SSL errors and not delivering valid responses
- was incorrectly configured to apply compression (gzip or brotli) to binary resources such as images

Both of these situations caused unhandled exceptions in the middleware, which logged untraceable errors.

### Fix upward-js ProxyResolver `ignoreSSLErrors` option

The ProxyResolver `ignoreSSLErrors` option should never be used in production, but it is an often useful workaround in development and staging, where SSL certificates are frequently invalid because of expiration dates, alternative names, or other small issues. The `upward-js` ProxyResolver `ignoreSSLErrors` implementation would ignore unrecognized certificate errors, but not any other kinds of SSL errors--such as a mismatched alt name due to proxied host headers. This PR fixes that by using a specially configured Node `httpsAgent`.

### Update Hastily to v0.4.6

Hastily would crash when trying to optimize a gzipped image stream. It's against best practices to configure a web server to gzip images, so rather than supporting it, Hastily now bypasses gzipped images and logs an informative error message. The real solution to #2448 is to reconfigure nginx (or whatever your web server uses) not to gzip images or other binary files:

From https://www.itworld.com/article/2693941/why-it-doesn-t-make-sense-to-gzip-all-content-from-your-web-server.html 
>You should not allow your web server to compress image files or PDF files. The reason being that these files are already compressed and by compressing them again you’re not only wasting CPU resources but you can actually make the resulting file larger by compressing them again.

From https://www.fastly.com/blog/new-gzip-settings-and-deciding-what-compress :
>When it comes to images, it’s generally assumed that content types starting with image/ should never be compressed.
 
From https://betterexplained.com/articles/how-to-optimize-your-site-with-gzip-compression/ :
>Most images, music and videos are already compressed. Don’t waste time compressing them again. In fact, you probably only need to compress the “big 3” (HTML, CSS and Javascript).

## Related Issue





Closes #2448.

## Acceptance 



### Verification Stakeholders
@dpatil-magento @Jordaneisenburger 
### Specification



### Verification Steps


1. Check out and `yarn install`.
2. Configure a local Magento instance to gzip image responses:
    1. If using Apache, add instructions to `.htaccess` in the Magento root. If using valet-plus, [here are instructions for finding nginx.conf](https://github.com/laravel/valet/issues/284).
    1. Use [the directives described here](https://varvy.com/pagespeed/enable-compression.html), making sure to add all MIME types for images. 
3. Run `yarn build && yarn stage:venia` and watch the logs.
4. Attempt to load the home page. Observe that hastily is logging error messages about the gzipped images, and the images themselves are coming through un-optimized.

## Checklist

* I have updated the documentation accordingly, if necessary.
